### PR TITLE
RCIAM-343_getCous_query_in_attrauthcomanage_is_not_working_properly

### PR DESCRIPTION
**Pull request by ioigoume**
- Fetch COPerson memberships in one query.
- Fix wrong calculation of entitlements if the user is an admin but has no affiliation in a COU.

### CODE REVIEW

**Comment by ioigoume** [comment]
@NicolasLiampotis please review

File: lib/Auth/Process/COmanageDbClient.php [hunk 1]
```diff
@@ -616,332 +583,537 @@ private function getGroups($personId, $coId)
...
+     * @todo Replace $co_id with a configuration variable
+     */
+    private function groupEntitlemeAssemble(&$state, $memberships_groups, $co_id, $voPrefix) {
+        if(empty($memberships_groups)) {
+            return;
+        }
+        foreach ($memberships_groups as $group) {
             $roles = array();
-            if ($group['member'] === true) {
+            if($group['member'] === true) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:1070]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 2]
```diff
@@ -616,332 +583,537 @@ private function getGroups($personId, $coId)
...
                 $roles[] = "member";
             }
-            if ($group['owner'] === true) {
+            if($group['owner'] === true) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:1073]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 3]
```diff
@@ -616,332 +583,537 @@ private function getGroups($personId, $coId)
...
                 $roles[] = "owner";
             }
-            if (!array_key_exists('eduPersonEntitlement', $state['Attributes'])) {
+            if(!array_key_exists('eduPersonEntitlement', $state['Attributes'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:1076]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 4]
```diff
@@ -616,332 +583,537 @@ private function getGroups($personId, $coId)
...
                     $this->urnNamespace          // URN namespace
                     . ":group:registry:"         // URN namespace
-                    . urlencode($groupName)      // VO
-                    . ":role=".$role             // role
+                    . urlencode($group['group_name'])      // VO
+                    . ":role=" . $role             // role
                     . "#" . $this->urnAuthority; // AA FQDN
                 // Enable legacy URN syntax for compatibility reasons?
-                if ($this->urnLegacy) {
+                if($this->urnLegacy) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:1092]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 5]
```diff
@@ -950,17 +1122,16 @@ private function getCouRootParent($cou_name, $cou_nested) {
...
+            }
         }
-      }
-      return false;
+        return false;
     }
 
-    private function _showException($e)
-    {
+    private function _showException($e) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:1134]: Opening braces for methods MUST go on the next line, and closing braces MUST go on the next line after the body. (see https://www.php-fig.org/psr/psr-2/#1-overview)

File: lib/Auth/Process/COmanageDbClient.php [hunk 6]
```diff
@@ -163,77 +164,32 @@ class sspmod_attrauthcomanage_Auth_Process_COmanageDbClient extends SimpleSAML_A
...
-        . ' AND NOT gr.deleted'
-        . ' AND gr.co_id = :coId'
-        . ' AND gr.group_type = \'S\''
-        . ' AND gr.status = \'A\''
-        . ' ORDER BY'
-        . ' gr.name DESC';
-
-    public function __construct($config, $reserved)
-    {
+    public function __construct($config, $reserved) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:167]: Opening braces for methods MUST go on the next line, and closing braces MUST go on the next line after the body. (see https://www.php-fig.org/psr/psr-2/#1-overview)

File: lib/Auth/Process/COmanageDbClient.php [hunk 7]
```diff
@@ -163,77 +164,32 @@ class sspmod_attrauthcomanage_Auth_Process_COmanageDbClient extends SimpleSAML_A
...
         parent::__construct($config, $reserved);
         assert('is_array($config)');
 
-        if (!array_key_exists('coId', $config)) {
+        if(!array_key_exists('coId', $config)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:171]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 8]
```diff
@@ -163,77 +164,32 @@ class sspmod_attrauthcomanage_Auth_Process_COmanageDbClient extends SimpleSAML_A
...
             SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'coId' not specified");
             throw new SimpleSAML_Error_Exception(
                 "attrauthcomanage configuration error: 'coId' not specified");
         }
-        if (!is_int($config['coId'])) {
+        if(!is_int($config['coId'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:176]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 9]
```diff
@@ -163,77 +164,32 @@ class sspmod_attrauthcomanage_Auth_Process_COmanageDbClient extends SimpleSAML_A
...
             SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'coId' not an integer number");
             throw new SimpleSAML_Error_Exception(
                 "attrauthcomanage configuration error: 'coId' not an integer number");
         }
         $this->coId = $config['coId'];
 
         // urnNamespace config
-        if (!array_key_exists('urnNamespace', $config) && !is_string($config['urnNamespace'])) {
+        if(!array_key_exists('urnNamespace', $config) && !is_string($config['urnNamespace'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:184]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 10]
```diff
@@ -163,77 +164,32 @@ class sspmod_attrauthcomanage_Auth_Process_COmanageDbClient extends SimpleSAML_A
...
           SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'urnNamespace' not specified or wrong format(string required)");
           throw new SimpleSAML_Error_Exception(
             "attrauthcomanage configuration error: 'urnNamespace' not specified");
         }
         $this->urnNamespace = $config['urnNamespace'];
 
         // voRoles config
-        if (!array_key_exists('voRoles', $config) && !is_string($config['voRoles'])) {
+        if(!array_key_exists('voRoles', $config) && !is_string($config['voRoles'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:192]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 11]
```diff
@@ -244,15 +200,15 @@ public function __construct($config, $reserved)
         $this->voRolesDef = $voRolesObject->getArrayCopy();
 
         // urnAuthority config
-        if (!array_key_exists('urnAuthority', $config) && !is_string($config['urnAuthority'])) {
+        if(!array_key_exists('urnAuthority', $config) && !is_string($config['urnAuthority'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:203]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 12]
```diff
@@ -244,15 +200,15 @@ public function __construct($config, $reserved)
...
           SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'urnAuthority' not specified or wrong format(string required)");
           throw new SimpleSAML_Error_Exception(
             "attrauthcomanage configuration error: 'urnAuthority' not specified");
         }
         $this->urnAuthority = $config['urnAuthority'];
 
         // Redirect Urls config
-        if (!array_key_exists('registryUrls', $config)) {
+        if(!array_key_exists('registryUrls', $config)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:211]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 13]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
           $invalid_urls = array_filter($config['registryUrls'], function($value) {
             return !filter_var($value, FILTER_VALIDATE_URL);
           });
-          if (!empty($invalid_keys) || !empty($invalid_urls)) {
+          if(!empty($invalid_keys) || !empty($invalid_urls)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:224]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 14]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
             SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'registryUrls' url or key configuration error");
             throw new SimpleSAML_Error_Exception(
               "attrauthcomanage configuration error: 'registryUrls' url or key configuration error");
           }
         }
         $this->registryUrls = $config['registryUrls'];
 
-        if (array_key_exists('coUserIdType', $config)) {
-            if (!is_string($config['coUserIdType'])) {
+        if(array_key_exists('coUserIdType', $config)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:232]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 15]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
+            if(!is_string($config['coUserIdType'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:233]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 16]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
                 SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'coUserIdType' not a string literal");
                 throw new SimpleSAML_Error_Exception(
                     "attrauthcomanage configuration error: 'coUserIdType' not a string literal");
             }
             $this->coUserIdType = $config['coUserIdType'];
         }
 
-        if (array_key_exists('userIdAttribute', $config)) {
-            if (!is_string($config['userIdAttribute'])) {
+        if(array_key_exists('userIdAttribute', $config)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:241]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 17]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
+            if(!is_string($config['userIdAttribute'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:242]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 18]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
                 throw new SimpleSAML_Error_Exception(
                     "attrauthcomanage configuration error: 'userIdAttribute' not a string literal");
             }
             $this->userIdAttribute = $config['userIdAttribute'];
         }
 
-        if (array_key_exists('blacklist', $config)) {
-            if (!is_array($config['blacklist'])) {
+        if(array_key_exists('blacklist', $config)) {
+            if(!is_array($config['blacklist'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:251]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 19]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
                 SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'blacklist' not an array");
                 throw new SimpleSAML_Error_Exception(
                     "attrauthcomanage configuration error: 'blacklist' not an array");
             }
             $this->blacklist = $config['blacklist'];
         }
-        if (array_key_exists('voWhitelist', $config)) {
-            if (!is_array($config['voWhitelist'])) {
+        if(array_key_exists('voWhitelist', $config)) {
+            if(!is_array($config['voWhitelist'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:259]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 20]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
                 SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'voWhitelist' not an array");
                 throw new SimpleSAML_Error_Exception(
                     "attrauthcomanage configuration error: 'voWhitelist' not an array");
             }
             $this->voWhitelist = $config['voWhitelist'];
         }
-        if (array_key_exists('communityIdps', $config)) {
-            if (!is_array($config['communityIdps'])) {
+        if(array_key_exists('communityIdps', $config)) {
+            if(!is_array($config['communityIdps'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:267]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 21]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
                 throw new SimpleSAML_Error_Exception(
                     "attrauthcomanage configuration error: 'communityIdps' not an array");
             }
             $this->communityIdps = $config['communityIdps'];
         }
 
-        if (array_key_exists('urnLegacy', $config)) {
-            if (!is_bool($config['urnLegacy'])) {
+        if(array_key_exists('urnLegacy', $config)) {
+            if(!is_bool($config['urnLegacy'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:276]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 22]
```diff
@@ -265,67 +221,67 @@ public function __construct($config, $reserved)
...
                 SimpleSAML_Logger::error("[attrauthcomanage] Configuration error: 'urnLegacy' not a boolean");
                 throw new SimpleSAML_Error_Exception(
                     "attrauthcomanage configuration error: 'urnLegacy' not a boolean");
             }
             $this->urnLegacy = $config['urnLegacy'];
         }
         if(array_key_exists('mergeEntitlements', $config)) {
-          if (!is_bool($config['urnLegacy'])) {
+          if(!is_bool($config['urnLegacy'])) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:284]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 23]
```diff
@@ -334,15 +290,14 @@ public function __construct($config, $reserved)
         }
     }
 
-    public function process(&$state)
-    {
+    public function process(&$state) {
         try {
             assert('is_array($state)');
-            if (isset($state['SPMetadata']['entityid']) && in_array($state['SPMetadata']['entityid'], $this->blacklist, true)) {
+            if(isset($state['SPMetadata']['entityid']) && in_array($state['SPMetadata']['entityid'], $this->blacklist, true)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:296]: https://www.php-fig.org/psr/psr-2/#51-if-elseif-else

File: lib/Auth/Process/COmanageDbClient.php [hunk 24]
```diff
@@ -334,15 +290,14 @@ public function __construct($config, $reserved)
         }
     }
 
-    public function process(&$state)
-    {
+    public function process(&$state) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/COmanageDbClient.php:293]: Opening braces for methods MUST go on the next line, and closing braces MUST go on the next line after the body. (see https://www.php-fig.org/psr/psr-2/#1-overview)

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]
Need to revert changes that break PSR-2 (or its newer PSR-12) which is part of the [contributing guidelines](https://github.com/simplesamlphp/simplesamlphp/blob/master/CONTRIBUTING.md)

**Comment by ioigoume** [comment]
Done with refactoring

File: lib/Auth/Process/COmanageDbClient.php (hunk 23 above)
Comment (ioigoume) [reply lib/Auth/Process/COmanageDbClient.php]: Done

File: lib/Auth/Process/COmanageDbClient.php (hunk 22 above)
Comment (ioigoume) [reply lib/Auth/Process/COmanageDbClient.php]: Done

**Review by NicolasLiampotis — APPROVED** [review]
LGTM

**Pull request merged**